src/gatt_connection.rs

```diff
@@ -0,0 +1,56 @@
+use crate::att_error::AttError;
+use core::fmt::Debug;
+
+use crate::bluetooth_address::BluetoothAddress;
+
+/// Represents a GATT connection from the peripheral perspective, specifically used as the API
+/// to allow the BLE peripheral device to respond to connected clients.
+pub trait GattConnection {
+  type SystemError: Debug;
+  type Responder: GattResponder + Debug;
+  type Writer: GattWriter + Debug + Clone;
```

Comment:
Oops, I think in hindsight that this can be achieved by just passing the Writer as an owned type in the event callback.  Responder is not owned on purpose because you aren't supposed to carry it with you to some other context (you can use it only during the event callback).  That said, I think for the implementation to work for complex threaded environments I might need to require Send on Writer... hmm, let me think about this a bit more and play with it in code.

Reply:
Yes, with or without an an owned `Writer` it needs to be `Send`.

Passing an owned instance of the writer to the callback might work, but you still have to make sure it cannot live longer than the `Peripheral` itself. And perhaps not even longer than the connection itself, if the connection becomes a non-simplistic handle object. So you need `Writer` to actually be, hmmm, `Writer<'a>` and then constraint `'a` either by the `GattConnection` lifetime, or by `Peripheral` lifetime - whichever is more correct.

Maybe let's clarify these two things first:
* Can a writer outlive its connection? Perhaps not, because I assume a writer is a writer for a particular connection, no?
* If above ^^^ is correct, why do you even bother sending a writer in the callback? It is enough to send a connection (owned one). The connection should of course not outlive the Peripheral and should be an owned object. Then the user can create a writer out from the connection at any time?